### ISSUE TITLE
adapter/compute/storage: cardinality statistics frequently timeout

### ISSUE THREAD

**Issue by mgree** [MaterializeInc/materialize#26986]
### What version of Materialize are you using?

v0.97

### What is the issue?

The [`CachedStatsOracle`](https://github.com/MaterializeInc/materialize/blob/d1427f1599848f9c3844d77dd74ba398a4b5700d/src/adapter/src/coord/sequencer/inner.rs#L4238-L4242) often works locally, but times out even with [substantially higher timeouts](https://github.com/MaterializeInc/materialize/blob/d1427f1599848f9c3844d77dd74ba398a4b5700d/src/adapter/src/coord/sequencer/inner.rs#L4229C1-L4236C11).

Stats are [currently ignored in join implementation](https://github.com/MaterializeInc/materialize/blob/d1427f1599848f9c3844d77dd74ba398a4b5700d/src/transform/src/join_implementation.rs#L274), so this is not high impact.

A few thoughts on what could be causing the problem here, from discussion with @bkirwi:

- timestamp selection (we use [`query_as_of`](https://github.com/MaterializeInc/materialize/blob/d1427f1599848f9c3844d77dd74ba398a4b5700d/src/adapter/src/coord/sequencer/inner/peek.rs#L561-L569), which may be too early) and/or interaction between strict serializability and timestamp selection
- actually slow communication around stats collection
- programmer error of async/timeouts

**Comment by bkirwi** [comment]
To expand on a couple of things:
- In general it's extremely possible for the query to select a timestamp that's not available in the sources yet, especially under strict serializability. That means the query will block -- which is fine, but we'd prefer the optimizer didn't.
    - OTOH, we probably don't want the stats at the latest available frontier, incase the sources are way ahead.
    - One possibility we discussed is using the query_as_of as an upper bound.
- It sounded like we were seeing timeouts even when the timeout was set at multiple seconds... which seems too high for the above to explain.

**Comment by maddyblue** [comment]
If the query timestamp will block, and this is causing the stats optimizer to timeout, there's no benefit (to the user) to choosing a lower timestamp for the stats because the query is still going to block (it'll just block in the dataflow instead of in the optimizer).

If we still want stats optimization to return quickly: what are the constraints or goals the stats optimizer has? From those I can recommend a good, fast way to get a timestamp.

**Comment by mgree** [comment]
> If the query timestamp will block, and this is causing the stats optimizer to timeout, there's no benefit (to the user) to choosing a lower timestamp for the stats because the query is still going to block (it'll just block in the dataflow instead of in the optimizer).

I don't totally follow this. If the timestamp blocks and takes longer than the query timeout, then we'll plan the query without stats. If we return stats within the timestamp, we'll plan the query with stats---and the query may block, but it may also be a better query (because our optimizer had more information).

> If we still want stats optimization to return quickly: what are the constraints or goals the stats optimizer has? From those I can recommend a good, fast way to get a timestamp.

I'd say the high-level desiderata are:

1. Get an estimate of the size of each source/dependency in the query's `source_ids`. The closer this estimate can come to "amount of diffs we'll have to process for that source", the better an estimate it is of the work done there.
2. Get that estimate quickly, so that we don't spend a long time collecting the stats. Low latency stats collection matters more for oneshot queries.

There are some semantic nits to pick about (1). We're currently using stats as a point-in-time estimate, but that doesn't make much sense for long-lived views and indices. I can imagine asking for the maximum number of diffs observed at any time (or some high percentile) to get an estimate of "heavy load we should plan for".

**Comment by bkirwi** [comment]
If 2 is a requirement (ie. we need a strict cap how much time is spent blocking on stats in the optimizer) then it doesn't really matter what timestamp we choose - any timestamp would have the possibility to block. So we can't seem to avoid the problem with timestamp selection alone.

That suggests to me that the timestamp-selection question is sort of orthogonal... we can choose whatever timestamp we'd otherwise choose for a query, and change the Persist API to return the closest available result without blocking.

**Comment by maddyblue** [comment]
> Get an estimate of the size of each source/dependency in the query's source_ids

Are the timestamp requirements for this the same as regular queries? The timestamp must be > the since of all source_ids and hopefully < the upper of all source ids?

> any timestamp would have the possibility to block

True, although we can do things to decrease the chance this happens. For example today most timestamps are chosen using the oracle read ts which is often hundreds of ms ahead of sources, causing regular delays during peeks (in strict serializable). We could have the timestamp chosen for stats use the same algorithm as serializable: ignore the oracle ts (and anything else contributing to lineraizability), and calculate the best timestamp we can using only sinces and uppers from the involved ids.

**Comment by maddyblue** [comment]
See https://github.com/MaterializeInc/materialize/pull/27555 for an implementation of this idea.

**Comment by mgree** [comment]
> See #27555 for an implementation of this idea.

Thanks, this is helpful! I can try to play around with this and see if I'm able to get meaningful stats on queries this way.

I think it might pay to talk a bit at a higher level about what we think these stats ought to mean. Right now we're using them to optimize peeks, which is not the most interesting case---long-lived indices and views are. It's not clear to me what stats should mean in the long-lived setting: when planning a join, do we want to know the high-water mark for a given input? Some kind of CDF of rows for a given input? Something else?

**Comment by bkirwi** [comment]
An observation we talked about in person, but I don't think is written down anywheres:

Persist stores it's data in batches, and older batches are larger than newer ones. So if for example we had a shard which covered times `[0, 16)`, we might have a batch from `[0, 8)`, then `[8, 12)`, then `[12, 14)`, etc. And on every batch boundary we can produce a decent size estimate: for example, we can estimate how many updates there are at time 12 by adding up the number of updates in the first two batches.

You could imagine fitting a function to these points, and getting an estimate of eg. the rate at which the size was growing, or try and estimate the max size over the next week or whatever. It's a little subtle to do correctly, since it's affected by eg. compaction. But we do have some access to historical data thanks to the shard's internal structure.

**Comment by mgree** [comment]
Oh, interesting! Should I understand these batches as the consolidated updates for those timestamps? Is there some boundary on the size of a batch or the amount of time it covers?

The main place we can use these statistics is in join planning, though there are others (see [the MIR cost model design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20230512_mir_cost_model.md#future-work)).

Differential joins are implemented as a single pipeline with arrangements of intermediate results. When an input updates, we only need to process inputs from that point in the pipeline on. So we might expect lower latency in joins where frequently updated inputs come later in the pipeline. We might expect lower memory usage in joins where inputs that need large arrangements come later in the pipeline (because we won't have to hold onto large intermediate results.)

Delta joins are implemented as n pipelines for n inputs, where each pipeline begins with a different input and there are arrangements for each input (but not for intermediate results). When an input updates, we run just that input's pipeline. It's less clear to me how stats could help here, but @frankmcsherry has done a bit of work on optimizing delta joins---he may have some tricks up his sleeve.

**Comment by bkirwi** [comment]
> Should I understand these batches as the consolidated updates for those timestamps?

Yes, though exactly how consolidated they are will vary - depends on how the since frontier has advanced etc.

> Is there some boundary on the size of a batch or the amount of time it covers?

Nope! In practice there is typically one initial batch, running from the beginning of time to something after when the shard was created (and can range from very large or empty, depending on the dataset). And then another batch that covers about half the time remaining, and then the next batch covers about half the time after that, etc... 

All very rough but hopefully that gives you some idea.

**Comment by mgree** [comment]
That makes sense, thank you! In the near term, I think approximate stats are better than no stats---but in the long term we'll probably want something with stronger guarantees than this.

When I'm done with #27472, I'll take a look at timestamp selection (via #27555). Once that's sorted, I think we (read: me, maybe Frank, maybe one of y'all if you want) should spend some time thinking about what statistics we want to collect and how.

**Comment by bkirwi** [comment]
> I think approximate stats are better than no stats---but in the long term we'll probably want something with stronger guarantees than this.

Worth spending some time thinking on what guarantees we want and can feasibly get! Right now we seem to be on a path to a world where the stats are both approximate (caching; selecting a serializable timestamp when the query is strict-serializable) and sometimes missing (because blocking is still possible) - maybe that's the right tradeoff, but I don't have enough clarity on the goals to be confident either way.

> Once that's sorted, I think we [...] should spend some time thinking about what statistics we want to collect and how.

I have no special expertise or opinion on the "what", but happy to chat about the "how"...

**Comment by mgree** [comment]
I played around #27555 today ([my `stats-ts` branch](https://github.com/mgree/materialize/tree/stats-ts)) and saw some weird behavior.

I'm running the following code, which is adapted from the reproducer for the issue that turned off stats in the first place.

```sql
SET ENABLE_SESSION_CARDINALITY_ESTIMATES TO TRUE;

DROP TABLE IF EXISTS tab0_raw, tab1_raw, tab2_raw, tab3_raw, tab4_raw CASCADE;

CREATE TABLE tab0_raw(pk INTEGER, col0 INTEGER, col1 FLOAT, col2 TEXT, col3 INTEGER, col4 FLOAT, col5 TEXT);
CREATE MATERIALIZED VIEW tab0 AS SELECT DISTINCT ON(pk) * FROM tab0_raw;
INSERT INTO tab0_raw VALUES(0,6,72.32,'diidw',65,65.1,'uudvn'), (1,57,90.1,'jvnyz',84,48.99,'raktj'), (2,68,91.83,'wefta',37,71.86,'zddoc'), (3,10,78.14,'zwjtc',7,9.96,'epmyn'), (4,63,24.41,'rwaus',66,53.7,'gbgmw'), (5,87,70.88,'rwpww',46,26.5,'bvbew'), (6,76,46.18,'lfvrf',99,92.47,'hqpgb'), (7,25,81.99,'khylz',54,73.22,'qaonp'), (8,93,17.58,'clxlk',88,59.16,'ziwhr'), (9,64,18.54,'fgkop',82,18.73,'lztum');

CREATE TABLE tab1_raw(pk INTEGER, col0 INTEGER, col1 FLOAT, col2 TEXT, col3 INTEGER, col4 FLOAT, col5 TEXT);
CREATE MATERIALIZED VIEW tab1 AS SELECT DISTINCT ON(pk) * FROM tab1_raw;

CREATE INDEX idx_tab1_0 on tab1 (col0);
CREATE INDEX idx_tab1_1 on tab1 (col1);
CREATE INDEX idx_tab1_3 on tab1 (col3);
CREATE INDEX idx_tab1_4 on tab1 (col4);

INSERT INTO tab1_raw SELECT * FROM tab0; -- SLOW????

CREATE TABLE tab2_raw(pk INTEGER, col0 INTEGER, col1 FLOAT, col2 TEXT, col3 INTEGER, col4 FLOAT, col5 TEXT);
CREATE MATERIALIZED VIEW tab2 AS SELECT DISTINCT ON(pk) * FROM tab2_raw;

CREATE INDEX idx_tab2_0 ON tab2 (col4);
CREATE INDEX idx_tab2_2 ON tab2 (col1);
CREATE INDEX idx_tab2_4 ON tab2 (col0,col3 DESC);
CREATE INDEX idx_tab2_5 ON tab2 (col3);

INSERT INTO tab2_raw SELECT * FROM tab0;

CREATE TABLE tab3_raw(pk INTEGER, col0 INTEGER, col1 FLOAT, col2 TEXT, col3 INTEGER, col4 FLOAT, col5 TEXT);
CREATE MATERIALIZED VIEW tab3 AS SELECT DISTINCT ON(pk) * FROM tab3_raw;

CREATE INDEX idx_tab3_0 ON tab3 (col0,col1,col3);
CREATE INDEX idx_tab3_1 ON tab3 (col4);

INSERT INTO tab3_raw SELECT * FROM tab0;

CREATE TABLE tab4_raw(pk INTEGER, col0 INTEGER, col1 FLOAT, col2 TEXT, col3 INTEGER, col4 FLOAT, col5 TEXT);
CREATE MATERIALIZED VIEW tab4 AS SELECT DISTINCT ON(pk) * FROM tab4_raw;

CREATE INDEX idx_tab4_0 ON tab4 (col4 DESC,col3 DESC);
CREATE INDEX idx_tab4_2 ON tab4 (col1);
INSERT INTO tab4_raw SELECT * FROM tab0;

EXPLAIN WITH(cardinality) SELECT pk FROM tab0 WHERE col0 <= 88 AND (((col4 <= 97.11 AND col0 IN (11,85,87,63,88) OR (col0 <= 45 AND ((((((((col0 > 79)) OR col1 <= 30.14 OR col3 >= 12))) OR col0 >= 89 OR col1 < 20.99 OR col1 >= 74.51 AND col3 > 77) AND (col0 IN (67,97,94,86,81))) AND ((col1 <= 10.70 AND col1 IS NULL AND col3 > 49 AND col3 > 66 AND (((col4 > 42.2) AND ((((col4 < 86.27) AND col3 >= 77 AND col3 < 48))) AND col3 >= 49)) AND col0 IN (SELECT col3 FROM tab0 WHERE col4 BETWEEN 20.3 AND 97.63))) AND col0 >= 25) OR ((col0 <= 35)) AND col0 < 68 OR ((col0 = 98))) OR (col1 <= 17.96) AND ((((col0 IS NULL))) OR col4 <= 2.63 AND (col0 > 2) AND col3 > 8) OR col3 <= 88 AND (((col0 IS NULL))) OR col0 >= 30)) AND col0 > 5) OR col0 > 3;

SELECT pk FROM tab0 WHERE col0 <= 88 AND (((col4 <= 97.11 AND col0 IN (11,85,87,63,88) OR (col0 <= 45 AND ((((((((col0 > 79)) OR col1 <= 30.14 OR col3 >= 12))) OR col0 >= 89 OR col1 < 20.99 OR col1 >= 74.51 AND col3 > 77) AND (col0 IN (67,97,94,86,81))) AND ((col1 <= 10.70 AND col1 IS NULL AND col3 > 49 AND col3 > 66 AND (((col4 > 42.2) AND ((((col4 < 86.27) AND col3 >= 77 AND col3 < 48))) AND col3 >= 49)) AND col0 IN (SELECT col3 FROM tab0 WHERE col4 BETWEEN 20.3 AND 97.63))) AND col0 >= 25) OR ((col0 <= 35)) AND col0 < 68 OR ((col0 = 98))) OR (col1 <= 17.96) AND ((((col0 IS NULL))) OR col4 <= 2.63 AND (col0 > 2) AND col3 > 8) OR col3 <= 88 AND (((col0 IS NULL))) OR col0 >= 30)) AND col0 > 5) OR col0 > 3;
```

1. I'm still missing stats at least some of the time with a 20ms timeout on my laptop.
2. When the system variable `enable_cardinality_estimiates` is on, I see run times in excess of 4min for the first `INSERT`, marked `SLOW???` in the excerpt above. If I turn it off it takes less than a minute. (That it's not instantaneous is surprising in its own right, but that could just be my very slow 2019 x86 macOS running mz in docker.)

**Comment by maddyblue** [comment]
Another thought: when the stats stuff wants to collect its data, it can wait for two reasons: 1) timestamp not yet ready and 2) actually fetching the data from disk/s3. The goal of #27555 is to reduce or eliminate only the delays caused by (1). But on our local machines (2) makes it hard to tell if this happened or not. I think what you want to measure is the output of `respond_immediately` method of `TimestampDetermination` on the stats ts: `stats_ts.unwrap().respond_immediately()` in that PR. (Well, really you want to not have unwrap and use the `ts` inside the `stats_ts` match. Recommend running this and only measuring  `respond_immediately` and not caring about the timeout because that's affected by local disk ops which we've seen be slow on macs.

**Comment by mgree** [comment]
Thank you for this advice!

Everything is reporting `respond_immediately() == true`. More worryingly, I'm unable to reproduce the slowness I saw two weeks ago. Still more worryingly, even with stats turned on I'm not seeing anything in `EXPLAIN` output.

I suspect either I'm holding something wrong or rebasing on main was a mistake. I'll look into it a bit more

**Comment by mgree** [comment]
Okay, I've figured out why we're not getting stats in the `EXPLAIN` output: the refactorings in the optimizer interface (and for `OptimizerTrace` in particular) aren't collecting any cardinality stats ([source](https://github.com/mgree/materialize/blob/stats-ts/src/adapter/src/explain/optimizer_trace.rs#L334)) and then `explain_peek` doesn't ever add any in ([source](https://github.com/mgree/materialize/blob/e7a126148e1d5adf00483b7c3fdc90fb3dc91270/src/adapter/src/coord/sequencer/inner/peek.rs#L216)).

Here's the state of play: 

- Ultimately, we should have cached stats on the controller that generally let us plan in a "good enough" way. (Maybe not for `AS OF` queries.)
- Right now, we collect stats pretty late in the query planning process.
- We need to either: (a) backpatch those stats into `OptimizerTrace` and its `ExplainContext`s ([usage 1](https://github.com/mgree/materialize/blob/e7a126148e1d5adf00483b7c3fdc90fb3dc91270/src/adapter/src/explain/optimizer_trace.rs#L356), [usage 2](https://github.com/mgree/materialize/blob/e7a126148e1d5adf00483b7c3fdc90fb3dc91270/src/adapter/src/explain/optimizer_trace.rs#L379)), or (b) get them early enough to create an `OptimizerTrace` with the stats already in there.

I'm not 100% clear on what the right fix is here; maybe @ggevay has an opinion?

**Comment by mgree** [comment]
Yes, that seems plausible! I'll try that this afternoon. Thanks! 😁
